src/modules/sensors/camera.c

```diff
@@ -290,18 +309,55 @@ static void recv_frame(int i) {
     state.brightness[i] = compute_brightness(buf.bytesused) / CAMERA_ILL_MAX;
 }
 
+static void reset_histogram(void) {
+    state.hist.count = memset(state.hist.count, 0, sizeof(double));
+    state.hist.sum = memset(state.hist.sum, 0, sizeof(double));
+}
+
 static double compute_brightness(const unsigned int size) {
     double brightness = 0.0;
+    double min = CAMERA_ILL_MAX;
+    double max = 0.0;
+
     /*
      * If greyscale -> increment by 1. 
      * If YUYV -> increment by 2: we only want Y! 
      */
     const int inc = 1 + (state.pixelformat == V4L2_PIX_FMT_YUYV);
-    
+
+    // Find minimum and maximum brightness
+    for (int i = 0; i < size; i += inc) {
+        if (state.buf.start[i] < min) {
+            min = state.buf.start[i];
+        }
+        if (state.buf.start[i] > max) {
+            max = state.buf.start[i];
+        }
+    }
+    if (max == 0.0) {
+        return brightness;
+    }
+
+    // Calculate histogram
+    const double step_size = (max - min) / HISTOGRAM_STEPS;
     for (int i = 0; i < size; i += inc) {
-        brightness += state.buf.start[i];
+        int bucket = (state.buf.start[i] - min) / step_size;
+        if (bucket >= 0 && bucket < HISTOGRAM_STEPS) {
+            state.hist.sum[bucket] += state.buf.start[i];
+            state.hist.count[bucket]++;
+        }
     }
-    brightness /= state.width * state.height;
+
+    // Default to the average for the highest brightness bucket
+    brightness = state.hist.sum[HISTOGRAM_STEPS-1] / state.hist.count[HISTOGRAM_STEPS-1];
+    // Find the highest brightness bucket that has a significant sample count and return the average brightness for that bucket
+    for (int i = HISTOGRAM_STEPS-1; i >= 0; i--) {
+        if (state.hist.count[i] > step_size) {
```

Comment:
Here we filter out buckets at the top that contain a low number of samples - I had some measurements where a point source of light reflecting off something would produce a handful of pixels at really high brightness. Using `step_size` here is entirely arbitrary - this could also be a fixed number, or some other scale, but using the step size seems to make the filter scale moderately well based on the granularity of the histogram.

Reply:
What about using a different filter, ie:
1) buckets with a good nice of samples
2) bucket with lowest variance

Second point should give us a bucket with homogeneous values that should produce better results.

Reply:
Part of the problem we have is that trying to determine ambient brightness from an image is simply not possible without knowing exactly how a particular camera sensor functions at a specific exposure, and even then I suspect it's error-prone.

1. I think you mean taking the highest population bucket? The problem is that images with large dark areas (due to furniture or similar) still won't return good results.
2. A bucket with low variance _and_ a low sample count is probably not representative though, and by their nature, there's a limit on possible variance imposed by the width of a bucket (which varies based on the total range of the scene).  The scenario I saw in testing, that prompted me to add a filter, was a bucket at the top that contained few samples, that were all quite close to each other, but caused by a small source of bright light, far in excess of anything else in the scene (there were a number of empty buckets below it, before the next substantial count).

Since we have the histogram now, we can probably roughly calculate the interquartile range and use that to eliminate outliers, which is the common statistical method - I'll try to take a look at that, and taking some test measurements, over the weekend.

Reply:
Actually I meant using both your current filter AND the second one: first found bucket that satisfies both of them.
This way we should use the highest bucket with a minimum number of samples and with lowest variance.

Obviously there is no "correct" approach, and i really like current algorithm's simpleness. That is why i proposed a small improvement on it :)

Reply:
I don't think variance is a good metric for filtering, because it's entirely random how much variance a bucket may contain, depending on where the bucket start/end happens to fall.

I've gone ahead and added some approximate calculations of the interquartile range, and used the general stats model of 1.5*IQR to filter any outlying buckets from the top and bottom of the histogram - this should provide pretty solid results, but I need to do some testing to make sure everything looks good now.

Reply:
That's great, thanks!